### ISSUE TITLE
SYS account connection state does not reflect cluster state

### ISSUE THREAD

**Issue by mullerch** [nats-io/nats-server#2703]
#### Versions of `nats-server` and affected client libraries used:

nats 2.6.4 (docker)

#### Steps or code to reproduce the issue:

Generate an account A with a user U from an operator (with a system user).

1. Start two servers, S1 and S2, with decentralized account auth, and connect them together in a single cluster
2. Subscribe $SYS.ACCOUNT.A.> on S1 with the system account
3. Connect a ~third server to S2 as a leafnode~ nats cli client to S2 with A
4. Send a request on $SYS.REQ.ACCOUNT.A.CONNZ on S1

#### Expected result:

The account connection state reports 1 connection.

#### Actual result:

The account connection state reports 0 connection.

If done on S2, the state reports 1 connection.

Note: step 2 is just here to show that the events are correctly managed, as there is a client_connect/client_disconnect event on the subscription done on S1 when the ~leafnode~client connects to S2

### Configurations

S1 config
```
server_name: broker-1

debug: true
#trace: true
port: 4222
monitor_port: 8222

cluster {
  name: cluster1
  port: 6222
  routes = [
    nats-route://broker-2:6222
  ]
}

jetstream {
  store_dir=nats
}

leafnodes {
  port: 7422
}

websocket {
  port: 443
  no_tls: true
}

mqtt {
  port: 1883
}

operator: "/nats/config/op.jwt"

resolver {
    type: full
    dir: '/nats/accounts'
    allow_delete: false
    interval: "2m"
}
```

S2 config (nearly identical)
```
server_name: broker-2

debug: true
#trace: true
port: 4222
monitor_port: 8222

cluster {
  name: cluster1
  port: 6222
  routes = [
    nats-route://broker-1:6222
  ]
}

jetstream {
  store_dir=nats
}

leafnodes {
  port: 7422
}

websocket {
  port: 443
  no_tls: true
}

mqtt {
  port: 1883
}

operator: "/nats/config/op.jwt"

resolver {
    type: full
    dir: '/nats/accounts'
    allow_delete: false
    interval: "2m"
}
```

**Comment by ripienaar** [comment]
Can you show your configs? When connecting a leafnode and wanting to see it you also have to add a remote for the system account.

**Comment by mullerch** [comment]
Updated with configs.

> When connecting a leafnode and wanting to see it you also have to add a remote for the system account.

I don't understand. What does "see my leafnode" mean? And where should I add a remote and to what?

**Comment by ripienaar** [comment]
Specifically keen to see the 3rd server config.

A route/gateway connects 2 servers at a low level where all communications in all accounts can flow. So when you ask a system account what it's aware of in those cases it's aware of all things that connects to it.

Leafnodes are different, leafnodes expand a specific account only, when you connect account FOO witha leafnode remote then thats the only account that will see traffic traverse the leafnode.  So the system accounts are not aware of each other hence the server doesnt show up and things like internal connection metrics are not shared.  This is on purpose to create a mechanism where 2 independent networks can safely connect without having to share internal data. 

You can add a 2nd remote to the leafnode to connect system accounts and then it might work - still, leafnodes are a bit different so dont expect full parity.

**Comment by mullerch** [comment]
Maybe there is a misunderstanding. I don't need the system account on the leafnode and the system account of the cluster to see each other. I need (or I expect to, related to what I've seen so far in nats) the system account to be consistent over the cluster (leafnode excluded), and its state not being dependent on the server the queries are made on.

I've connected a leafnode, but I guess this inconsistency could also be reproduced by connection a standard client to the cluster.

**Comment by mullerch** [comment]
I've just tested with a nats CLI client and it's the same behavior

**Comment by ripienaar** [comment]
OK, right I missunderstood you indeed.

The servers each report only their own view of connections that they know about, not cluster wide. Maintaining such a structure over large clusters would require a large amount of consistency protocol stuff which doesnt exist for this.

So if you look for example what ```nats server report accounts --context system --trace``` does you'll see it gets results from all servers and then calculate totals.

**Comment by mullerch** [comment]
Ok, I understand.

To me there is a little inconsistency between the events (eg.g $SYS.ACCOUNT.<id>.CONNECT) and the state requests responses (e.g. $SYS.REQ.ACCOUNT.<account-id>.CONNZ). As they have the same subject path, one may think they bring the same level of abstraction of the cluster, where it's not the case.

To make this more obvious and self-documenting, you may reconsider those paths. My suggestion:

- (Cluster specific) account information: $SYS.ACCOUNT.<id> (REQ would not exist as you said)
- Server specific account information: $SYS.(REQ.)SERVER.ACCOUNT.<id>

The same applies to other endpoints. That said, $SYS.SERVER.ACCOUNT.<id>.CONNS is already used to send cluster-wide information events (retro-compatibility issues)

Also, please note the [doc](https://docs.nats.io/nats-server/configuration/sys_accounts#available-events-and-services) is not in sync with the implementation at this time. 

> "$SYS.REQ.ACCOUNT.<account-id>.<endpoint-name> (from all server, request account specific monitoring endpoint corresponding to account id and endpoint name - will return multiple messages)

Should state "from the connecting server". By the way, I did never receive multiple messages so far.

There is a `$SYS.ACCOUNT.<id>.SERVER.CONNS` event that is not documented.

`$SYS.ACCOUNT.<id>.LEAFNODE.CONNECT` and `$SYS.ACCOUNT.<id>.LEAFNODE.DISCONNECT` do not exist anymore (see #2291).

The bottom line is that those monitoring features are really awesome and useful, but at this time is is quite confusing how it's done, and the documentation does not make it trivial either.

**Comment by ripienaar** [comment]
When you say you dont receive multiple messages, how do you request them? With nats req?

```
$ nats sub -i --context system
11:30:12 Subscribing on _INBOX.WzH01RQFK55Z0WKKlU8VN1
```

Followed by:

```
$ nats pub --context system '$SYS.REQ.ACCOUNT.A.CONNZ' '' --reply _INBOX.WzH01RQFK55Z0WKKlU8VN1
```

My subscriber will get 9 messages - 9 servers.

The request command is 1:1, 1 request, 1 reply stops.

**Comment by mullerch** [comment]
> The request command is 1:1, 1 request, 1 reply stops.

Oh I did not knew that. Know I understand how it works! And the subject names make more sense. There is an abstraction for the request, but not for the replies. 

This seems acceptable as we don't need to connect individually on each server as I thought.

Thanks for your help.

**Comment by ripienaar** [comment]
Cool, glad that makes more sense now, having a 1:n request-reply function is a often requested feature fwiw, but for now yeah thats why its confusing.

**Comment by ripienaar** [comment]
Also fwiw, `nats server req accounts` is probably what you wanted in the first place, its a 1:n request for this info and also understands how long to wait for etc :)

**Comment by mullerch** [comment]
I think it's rather `nats server report accounts <id>` I'm looking for. But thank for pointing out I'll check the source so I don't miss anything for my implementation. Btw it's reporting an error if there are no connections: `nats: error: received results for multiple accounts, try --help`
